components/script/serviceworkerjob.rs

```diff
@@ -147,30 +149,28 @@ impl JobQueue {
     }
     #[allow(unrooted_must_root)]
     // https://w3c.github.io/ServiceWorker/#schedule-job-algorithm
-    pub fn schedule_job(&mut self, job: Job,
-                           reg: &ServiceWorkerRegistration,
-                           script_thread: &ScriptThread) {
+    pub fn schedule_job(&mut self,
+                        job: Job,
+                        global: &GlobalScope,
+                        script_thread: &ScriptThread) {
+        // NOTE: need to get hold of promise here early on, as we are pushing the job into the job queue
         let promise = job.promise.clone();
-        if !self.0.contains_key(&job.scope_url) {
-            self.0.insert(job.scope_url.clone(), vec![]);
-        }
-        let ref mut job_queue = match self.0.get_mut(&job.scope_url) {
-            Some(r) => r,
-            None => return
-        };
+        let job_queue = self.0.entry(job.scope_url.clone()).or_insert(vec![]);
         // Step 1
         if job_queue.is_empty() {
             let scope_url = job.scope_url.clone();
             job_queue.push(job);
             let run_job_handler = RunJobHandler {
-                reg: Trusted::new(reg),
+                global: Trusted::new(global),
                 promise: TrustedPromise::new(promise),
                 scope_url: scope_url
             };
             // queue task for https://w3c.github.io/ServiceWorker/#run-job-algorithm
-            script_thread.queue_run_job(box run_job_handler, &*reg.global());
+            script_thread.queue_run_job(box run_job_handler, global);
         } else {
             // or Step 2
+            // below lines doesn't look very appropriate, but lifetimes forces me to do it this way.
+            // are there ways to get mutable ref to last_job of job_queue ?
```

Comment:
Still getting the same lifetime issue, even using that as there is a case of more than once mutable borrow here. So, should we let it stay this way ?